<h1>INSTANCE-BASED LEARNING</h1>
<h2>From the book: Machine Learning by Tom M. Mitchell page 230</h2>
<img src="images/inst.jpg">

<h1 title='K-Nearest Neighbour'>K-NN</h1>

The KNN is one of the simplest machine learning algorithms and an example of instance-based learning, where new instances are predicted based on stored instances. k-nearest neighbors algorithm (k-NN) is a supervised nonparametric method used for both classification (data with discrete labels) and regression (data with continuous labels) but  widely used for classification problems. At the heart of this method is the K ( The number of nearest neighbors to look for) parameter. Unlike other supervised learning algorithms, K-NN does not learn an explicit mapping f from the training data which will used in making predictions or classifying new instances but instead with a predefined K, KNN finds the K training samples that are closest to the new instance based some distance or similarity function such as Euclidean (commonly used), Manhattan, Mahalanobis, cosine similarity function and predict the label from these K closest training example.

<h1 title='K-NN USED AS A CLASSIFIER'>In classification settings</h1> 
Once k is selected, given a new instance the method assigns a label to a new instance (unlabeled vector) based on the k closest training data points,  that is a new instance is classified by assigning the label which is most frequent or common among the k training samples nearest to that new instance.

<h1 title='K-NN USED AS A REGRESSOR'>In regression </h1>
assign the average response (Average the values of k nearest neighbors)  of the of k nearest neighbors

<h1> K-Nearest Neighbors Algorithm </h1>
<ul>
    <li>Determine parameter K </li>
    <li>Compute the distance of the new instance from each training point </li>
   <li>Sort the distances in ascending (or descending) order </li>
<li>Select the K nearest neighbors based on the sorted distances </li>
<li>Use majority rule (for classification) or averaging (for regression)</li>
</ul>

<h1>NOTE</h1>
The choice of distance or similarity function depends on the type of features in the dataset

The experimental results from [Effects of Distance Measure Choice on KNN Classifier Performance - A Review](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4978658/) show that the performance of KNN classifier depends significantly on the distance measure used. The results also show that some distance measure are less affected by added noise.


In [1]:
from sklearn.datasets import load_iris
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
from sklearn.preprocessing import StandardScaler
import numpy as np

In [2]:
np.random.seed(1000)

In [3]:
iris=load_iris()

In [4]:
features=iris.data
labels=iris.target

In [5]:
features.shape,labels.shape

((150, 4), (150,))

In [6]:
scale=StandardScaler()
features=scale.fit_transform(features)    

In [7]:
features[1:6]

array([[-1.14301691, -0.13197948, -1.34022653, -1.3154443 ],
       [-1.38535265,  0.32841405, -1.39706395, -1.3154443 ],
       [-1.50652052,  0.09821729, -1.2833891 , -1.3154443 ],
       [-1.02184904,  1.24920112, -1.34022653, -1.3154443 ],
       [-0.53717756,  1.93979142, -1.16971425, -1.05217993]])

the data contain 150 instances with 4 features and a label

In [8]:
labels[1:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0])

# spliting the data into training and test dataset with 20% as test dataset

In [9]:
indices = np.random.permutation(len(features))
X_train = features[indices[:-30]]
y_train = labels[indices[:-30]]
X_test = features[indices[-30:]]
y_test = labels[indices[-30:]]

In [10]:
y_train[1:10]

array([0, 2, 2, 0, 0, 1, 1, 0, 2])

In [11]:
y_train.shape

(120,)

# instantiating a KNN OBJECT with k=6

In [12]:
knn=KNeighborsClassifier(n_neighbors=6)

In [13]:
knn.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=6, p=2,
                     weights='uniform')

<h1>Minkowski Distance</h1>
The Minkowski distance can be considered as a generalization of both the Euclidean and Manhattan distance.
The Minkowski distance of order p where p is an ineteger between two points
$X=(x_{1},x_{2},\cdots, x_{n})$ $Y=(y_{1},y_{2},\cdots, y_{n}$ is defined as
$$ D(X,Y)=\lgroup \sum_{i}^{n}|x_{i}-y_{i}|^{p}    \rgroup ^{\frac{1}{p} }$$

The default metric is minkowski with p=2 which is equivalent to the standard Euclidean distance.

<h1>Euclidean Distance</h1>
between two points X and Y is defined as

$$ D(X,Y)=\sqrt {\sum_{i}^{n}(x_{i}-y_{i})^{2}} $$
<h1>Manhattan Distance or $L^{1}$ distance OR $\ell_{1}$ norm</h1>
between two points X and Y is defined as
$$ D(X,Y)=\sum_{i}^{n}|x_{i}-y_{i}| $$

In [14]:
y_pred=knn.predict(X_test)
y_pred

array([0, 0, 2, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 2, 1, 0, 1, 0, 0, 2, 2, 0,
       0, 1, 2, 0, 1, 1, 1, 1])

In [15]:
y_test

array([0, 0, 2, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 2, 1, 0, 1, 0, 0, 2, 2, 0,
       0, 1, 2, 0, 1, 1, 1, 1])

In [16]:
import pandas as pd

In [17]:
y_pred==y_test

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True])

<h1>Performance Metrics</h1>
Since the problem is treated as a classification problem, like any other classification problem, F1−score,
precision and accuracy were the performance metrics used.
<h1>Accuracy</h1>
  is defined to be the ratio of correctly predicted observation to the total observations and it heuristically
provides us with information as whether a model is trained correctly or not and is defined as

$$Accuracy = \frac{T rue\ positive + T rue \ negative}{Total \ sample}$$

<h1>Precision</h1>
 tell us if the various labeled intents are correctly predicted positive by the classifier then
how often are they correctly predicted positive and is a good measure of determining the costs of false
positive predicted. Precision is defined as the ratio of correctly predicted positive observations to the
total predicted positive observations (both falsely and truly predicted positive). Precision is defined as

$$P recision =\frac{ True \ positive}{True \ positive + False \ positive}$$
<h1>F1−Score</h1>
is a measure of a test’s accuracy which can be interpreted as a weighted average of the
precision and recall reaching its best value at 1 and worst score at 0 and is defined as

$$ F1 − Score = 2 ∗ \frac{precision ∗ recall}{precision + recal }$$


<h1>Recall</h1>
 is defined as
$$recall = \frac{True \ positive}{True \ positive + False \ negative}$$

In [18]:
accuracy_score(y_test,y_pred)

1.0

In [19]:
confusion_matrix(y_test,y_pred)

array([[13,  0,  0],
       [ 0, 12,  0],
       [ 0,  0,  5]], dtype=int64)

In [20]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        13
           1       1.00      1.00      1.00        12
           2       1.00      1.00      1.00         5

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30

